Remove some overhead from Process.Start

stephentoub opened the pull request:
- Avoid StringBuilder marshaling
- Let CreateProcess{WithLogonW} determine the current working directory
- Avoid creating SafeHandles for GetCurrentProcess()
- Avoid forcing ProcessStartInfo.ArgumentList into existence just to check if it contains anything
- Prefer using ProcessStartInfo.Arguments in Start(string, IEnumerable<string>) if there's only one string in the enumerable
- Use ValueStringBuilder instead of StringBuilder to build up arguments, so as to use stack space / pooled char[]s and avoid actually needing to produce strings.
- Avoid unnecessary SafeHandle for PROCESS_INFORMATION.hThread

Test that launches a thousand processes, no redirection, one argument (execution time is dominated by the actual native call and remote process, so this has a negligible impact on serial execution time):

Allocations before
![image](https://user-images.githubusercontent.com/2642209/99155381-6f047100-2685-11eb-819d-1c7d8b0de018.png)

Allocations after
![image](https://user-images.githubusercontent.com/2642209/99155392-86435e80-2685-11eb-8a08-83849bf57ea2.png)

Contributes to #44598

jkotas commented on src/libraries/Common/src/Interop/Windows/Kernel32/Interop.DuplicateHandle_SafeProcessHandle.cs:
```diff
@@ -11,9 +11,9 @@ internal partial class Kernel32
     {
         [DllImport(Libraries.Kernel32, SetLastError = true, BestFitMapping = false)]
         internal static extern bool DuplicateHandle(
```
It would be nice to rename the .cs file now that it has nothing to do with SafeProcessHandle.

jkotas commented on src/libraries/System.Diagnostics.Process/src/System.Diagnostics.Process.csproj:
```diff
@@ -99,7 +101,7 @@
              Link="Common\Interop\Windows\Kernel32\Interop.TerminateProcess.cs" />
-    <Compile Include="$(CommonPath)Interop\Windows\Kernel32\Interop.GetCurrentProcess_SafeProcessHandle.cs"
+    <Compile Include="$(CommonPath)Interop\Windows\Kernel32\Interop.GetCurrentProcess_IntPtr.cs"
              Link="Common\Interop\Windows\kernel32\Interop.GetCurrentProcess.cs" />
```
```suggestion
             Link="Common\Interop\Windows\kernel32\Interop.GetCurrentProcess_IntPtr.cs" />
```

jkotas commented on src/libraries/System.Diagnostics.PerformanceCounter/src/System.Diagnostics.PerformanceCounter.csproj:
```diff
@@ -78,7 +78,7 @@
              Link="Common\Interop\Windows\Kernel32\Interop.GetComputerName.cs" />
-    <Compile Include="$(CommonPath)Interop\Windows\Kernel32\Interop.GetCurrentProcess_SafeProcessHandle.cs"
+    <Compile Include="$(CommonPath)Interop\Windows\Kernel32\Interop.GetCurrentProcess_IntPtr.cs"
              Link="Common\Interop\Windows\kernel32\Interop.GetCurrentProcess.cs" />
```
```suggestion
             Link="Common\Interop\Windows\kernel32\Interop.GetCurrentProcess_IntPtr.cs" />
```

jkotas commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:
```diff
@@ -1276,9 +1276,26 @@ public static Process Start(string fileName, IEnumerable<string> arguments)
+                    string firstArgument = e.Current;
+                    if (!e.MoveNext())
+                    {
+                        startInfo.Arguments = firstArgument;
```
Are we still going to get the escaping of special characters with this optimization?

stephentoub replied:
You're right.  I forgot we made Argument and ArgumentList behave differently. It's unintuitive we did that,  but at least it's doc'd; even more so,  and not doc'd, that there's an escaping difference between the `Start(..., string)` and `Start(..., IEnumerable<string>)`.  And it's not worth complicating tons further to add escaping here.  Will undo.

jkotas commented on src/libraries/System.Security.Principal.Windows/src/System.Security.Principal.Windows.csproj:
```diff
@@ -59,8 +59,8 @@
              Link="Common\Interop\Interop.DuplicateTokenEx.cs" />
-    <Compile Include="$(CommonPath)Interop\Windows\Kernel32\Interop.DuplicateHandle.cs"
-             Link="Common\Interop\Interop.DuplicateHandle.cs" />
+    <Compile Include="$(CommonPath)Interop\Windows\Kernel32\Interop.DuplicateHandle.SafeAccessTokenHandle.cs"
```
Nit: I believe that the usual naming convention under common/interop is to use underscore for the specialization, e.g.: `Interop.DuplicateHandle_SafeAccessTokenHandle.cs`

eiriktsarpalis commented on src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:
```diff
@@ -645,6 +648,8 @@ private unsafe bool StartWithCreateProcess(ProcessStartInfo startInfo)
                 _standardError = new StreamReader(new FileStream(parentErrorPipeHandle!, FileAccess.Read, 4096, false), enc, true, 4096);
             }
 
+            commandLine.Dispose();
```
Why not defined with a `using` statement?

stephentoub replied:
We've generally not added try/finallys just for returning buffers to the array pool.  I'd be fine either way, though.

eiriktsarpalis commented on src/libraries/System.Private.CoreLib/src/System/PasteArguments.Unix.cs:
```diff
@@ -14,10 +14,10 @@ internal static partial class PasteArguments
         internal static string Paste(IEnumerable<string> arguments, bool pasteFirstArgumentUsingArgV0Rules)
         {
-            var stringBuilder = new StringBuilder();
+            var stringBuilder = new ValueStringBuilder(stackalloc char[512]);
```
Is there a reason why we're allocating 512 instead of 256 on windows?

stephentoub replied:
Nope, I'd started with 512 and then backed off to 256 when I saw us already using 256 in the same project elsewhere, but neglected to revert this one.

eiriktsarpalis reviewed (approved):
LGTM.

eiriktsarpalis commented:
/azp run runtime-libraries outerloop

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/PasteArguments.Unix.cs:
```diff
@@ -14,10 +14,10 @@ internal static partial class PasteArguments
         internal static string Paste(IEnumerable<string> arguments, bool pasteFirstArgumentUsingArgV0Rules)
         {
-            var stringBuilder = new StringBuilder();
+            var stringBuilder = new ValueStringBuilder(stackalloc char[512]);
```
```suggestion
            var stringBuilder = new ValueStringBuilder(stackalloc char[256]);
```

jkotas commented on src/libraries/System.IO.Pipes/src/System.IO.Pipes.csproj:
```diff
@@ -49,10 +49,10 @@
              Link="Common\Interop\Windows\Interop.FileOperations.cs" />
     <Compile Include="$(CommonPath)Interop\Windows\Kernel32\Interop.FileTypes.cs"
              Link="Common\CoreLib\Interop\Windows\Interop.FileTypes.cs" />
```
```suggestion
             Link="Common\Interop\Windows\Interop.FileTypes.cs" />
```

jkotas commented on src/libraries/System.IO.Pipes/src/System.IO.Pipes.csproj:
```diff
@@ -49,10 +49,10 @@
+    <Compile Include="$(CommonPath)Interop\Windows\Kernel32\Interop.DuplicateHandle_SafePipeHandle.cs"
+             Link="Common\Interop\Windows\Interop.DuplicateHandle_SafePipeHandle.cs" />
     <Compile Include="$(CommonPath)Interop\Windows\Kernel32\Interop.GetFileType_SafeHandle.cs"
              Link="Common\CoreLib\Interop\Windows\Interop.GetFileType.cs" />
```
```suggestion
             Link="Common\Interop\Windows\Interop.GetFileType.cs" />
```

jkotas commented:
Build break
```
[F:\workspace\_work\1\s\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
Source file 'F:\workspace\_work\1\s\src\libraries\Common\src\Interop\Windows\Kernel32\Interop.GetCurrentProcess_IntPtr.cs' could not be found.
```

stephentoub committed:
Remove some overhead from Process.Start

- Avoid StringBuilder marshaling
- Let CreateProcess{WithLogonW} determine the current working directory
- Avoid creating SafeHandles for GetCurrentProcess()
- Avoid forcing ProcessStartInfo.ArgumentList into existence just to check if it contains anything
- Prefer using ProcessStartInfo.Arguments in Start(string, IEnumerable<string>) if there's only one string in the enumerable
- Use ValueStringBuilder instead of StringBuilder to build up arguments, so as to use stack space / pooled char[]s and avoid actually needing to produce strings.
- Avoid unnecessary SafeHandle for PROCESS_INFORMATION.hThread

stephentoub commented:
> Build break

What happens when you search within *.csproj and forget about *.projitems ;-)
